server/asset/eth/rpcclient.go

```diff
@@ -52,111 +56,268 @@ type ethConn struct {
+// If at least one of the connections is successful and is not outdated, the
+// function will return without error.
+//
+// Connections with an outdated block will be marked as outdated, but included
+// in the clients slice. If the up-to-date providers start to fail, the outdated
+// ones will be checked to see if they are still outdated.
+//
+// Failed connections will not be included in the clients slice.
 func (c *rpcclient) connect(ctx context.Context) (err error) {
-	netAddrs, found := dexeth.ContractAddresses[ethContractVersion]
-	if !found {
-		return fmt.Errorf("no contract address for eth version %d", ethContractVersion)
-	}
-	contractAddr, found := netAddrs[c.net]
-	if !found {
-		return fmt.Errorf("no contract address for eth version %d on %s", ethContractVersion, c.net)
-	}
-
 	var success bool
 
-	c.clients = make([]*ethConn, len(c.endpoints))
-	for i, endpoint := range c.endpoints {
-		client, err := rpc.DialContext(ctx, endpoint)
+	c.clients = make([]*ethConn, 0, len(c.endpoints))
+	c.neverConnectedEndpoints = make([]string, 0, len(c.endpoints))
+
+	for _, endpoint := range c.endpoints {
+		ec, err := c.connectToEndpoint(ctx, endpoint)
 		if err != nil {
-			return fmt.Errorf("unable to dial rpc to %q: %v", endpoint, err)
+			c.log.Errorf("Error connecting to %q: %v", endpoint, err)
+			c.neverConnectedEndpoints = append(c.neverConnectedEndpoints, endpoint)
+			continue
 		}
 
 		defer func() {
+			// If all connections are outdated, we will not start, so close any open connections.
 			if !success {
-				client.Close()
+				ec.Close()
 			}
 		}()
 
-		ec := &ethConn{
-			Client:   ethclient.NewClient(client),
-			endpoint: endpoint,
-			tokens:   make(map[uint32]*tokener),
-		}
+		c.clients = append(c.clients, ec)
+	}
 
-		reqModules := []string{"eth", "txpool"}
-		if err := dexeth.CheckAPIModules(client, endpoint, c.log, reqModules); err != nil {
-			c.log.Warnf("Error checking required modules at %q: %v", endpoint, err)
-			c.log.Warnf("Will not account for pending transactions in balance calculations at %q", endpoint)
-			ec.txPoolSupported = false
-		} else {
-			ec.txPoolSupported = true
-		}
+	success = c.sortConnectionsByHealth(ctx)
 
-		hdr, err := ec.HeaderByNumber(ctx, nil)
-		if err != nil {
-			return fmt.Errorf("error getting best header from %q: %v", endpoint, err)
-		}
-		if c.headerIsOutdated(hdr) {
-			return fmt.Errorf("initial header fetched from %q appears to be outdated (time %s). If you continue to see this message, you might need to check your system clock",
-				endpoint, time.Unix(int64(hdr.Time), 0))
-		}
+	if !success {
+		return fmt.Errorf("failed to connect to an up-to-date ethereum node")
+	}
 
-		es, err := swapv0.NewETHSwap(contractAddr, ec.Client)
-		if err != nil {
-			return fmt.Errorf("unable to initialize eth contract for %q: %v", endpoint, err)
-		}
-		ec.swapContract = &swapSourceV0{es}
-		ec.caller = client
+	go c.monitorConnectionsHealth(ctx)
```

Comment:
I think it's fine to have this unsupervised based on the caller's pattern of `connect`'s context cancellation followed by calling the `shutdown()` method, but it's a little confusing.  Seem like anything could go wrong?

Fundamentally I don't quite get why there's a there's both a context for connect, but also a shutdown method.  It's like the context on connect was sort of intended to just apply to the initial connection action (e.g. timeout), but this use of it makes it apply to longer running processes, which is compatible with how `(*ETHBackend).Connect` works...

Does this all seem fine or are things a bit mixed up?

Reply:
I got rid of the shutdown method. I guess the reason was to make sure that the ETH backend was shutdown before closing the client connections, but that can be handled with a new context as well.